src/analysis/retail/shaman/enhancement/modules/normalizers/EventOrderNormalizer.tsx

```diff
@@ -44,52 +67,111 @@ const primordialWaveEventOrder: EventOrder = {
    * from Windstrike into 'freecast' so they don't interfere with the APL */
   normalize(events: AnyEvent[]) {
     events = super.normalize(events);
 
     const fixedEvents: AnyEvent[] = [];
-    const thorimsInvocationCastIds = [SPELLS.LIGHTNING_BOLT.id, TALENTS.CHAIN_LIGHTNING_TALENT.id];
+    const thorimsInvocationCastIds = [
+      SPELLS.LIGHTNING_BOLT.id,
+      TALENTS.CHAIN_LIGHTNING_TALENT.id,
+      SPELLS.TEMPEST_CAST,
+    ];
     const windstrikeId = SPELLS.WINDSTRIKE_CAST.id;
+    const skipEvents = new Set<AnyEvent>();
 
     events.forEach((event: AnyEvent, idx: number) => {
-      fixedEvents.push(event);
+      if (!skipEvents.has(event)) {
+        fixedEvents.push(event);
+      }
       // non-cast events are irrelevant
       if (event.type !== EventType.Cast) {
         return;
       }
-      // only interested in Lightning Bolt and Chain Lightning
+
       const spellId = event.ability.guid;
-      if (!thorimsInvocationCastIds.includes(spellId)) {
-        return;
-      }
+      if (thorimsInvocationCastIds.includes(spellId)) {
+        // For ChL, LB, and Tempest casts, look backwards for a Windstrike cast
+        for (let backwardsIndex = idx - 1; backwardsIndex >= 0; backwardsIndex -= 1) {
+          const backwardsEvent = events[backwardsIndex];
+          // The windstrike and auto cast typically occur on the same timestamp
+          if (event.timestamp - backwardsEvent.timestamp > EventLinkBuffers.MaelstromWeapon) {
+            break;
+          }
 
-      // For ChL and LB casts, look backwards for a Windstrike cast
-      for (let backwardsIndex = idx - 1; backwardsIndex >= 0; backwardsIndex -= 1) {
-        const backwardsEvent = events[backwardsIndex];
-        // The windstrike and auto cast typically occur on the same timestamp
-        if (event.timestamp - backwardsEvent.timestamp > MAELSTROM_WEAPON_MS) {
-          break;
+          if (
+            backwardsEvent.type !== EventType.Cast ||
+            backwardsEvent.ability.guid !== windstrikeId
+          ) {
+            continue;
+          }
+
+          fixedEvents.splice(idx, 1);
+          fixedEvents.push({
+            ...event,
+            type: EventType.FreeCast,
+            __modified: true,
+          });
         }
+      }
 
-        if (
-          backwardsEvent.type !== EventType.Cast ||
-          backwardsEvent.ability.guid !== windstrikeId
-        ) {
-          continue;
+      /** This interaction is a bug. The maelstrom is consumed after the cast so doesn't increase damage, but does apply to
+       * Static Accumulation refund. */
+
+      /** The Rolling Thunder hero talent summons the feral spirt AFTER the tempest cast (but before damage). The issue
+       * with this is when a feral spirit is summoned, it also immediately generates 1 stack of msw, but because this
+       * appearing after, we need to move it back. a traditional order normalizer doesn't move the events correctly */
```

Comment:
is the issue with the normal approach that you have to do this only if you see all 3? (i see the `if (eventsToMoveBack.length === 3)` below)

because if so i think this is the correct way to go. that is so special that it probably isn't needed in core.

Reply:
Yeah that's right. If any of the 3 are missing, it's not a "match" and they shouldn't be reordered